templates/docs/examples/patterns/forms/_checkbox-heading.html

```diff
@@ -0,0 +1,39 @@
+<h1>
+  <div class="p-checkbox--heading">
+    <div class="p-checkbox__input">
+      <input type="checkbox" id="checkExamplePatternH1">
+      <label for="checkExamplePatternH1" />
+    </div>
+    <label class="p-checkbox__label" for="checkExamplePatternH1">.p-checkbox--heading</label>
+  </div>
+</h1>
+
+<h2>
+  <div class="p-checkbox--heading">
+    <div class="p-checkbox__input">
+      <input type="checkbox" id="checkExamplePatternH2">
+      <label for="checkExamplePatternH2" />
```

Comment:
No. We can't do it on parent, because parent is not aware of the checkbox state. It also has to be a label, because clicking on it needs to affect the checkbox itself.

The only way to do a styled checkbox is `input + label` structure. So technically speaking the label inside the `p-checkbox__input` is solely for styled checkbox. Then we need a second label for a text label on the checkbox. This label is optional in the component. It can be placed anywhere outside and pointed to the checkbox (for example when there is a grid and labels are in different column).

I know that this structure may be confusing at first. Maybe we should make sure it's better documented (to warn developers that these 2 labels are expected and they both need to point at the same checkbox input).

Material UI does something similar with checkboxes: https://material-ui.com/components/checkboxes/#checkbox Although in their case I guess you don't really build this markup yourself, it's rendered from a JS component.

@anthonydillon You have any thought on this?

Reply:
I this case why not wrap the input and the custom markup required in a label? Something like:
```html
<span class="p-checkbox">
  <label class="p-checkbox__label">
    <input type="checkbox" class="p-checkbox__input" />
    <span class="p-checkbox__label-text">.p-checkbox--heading</span>
  </label>
</span>
```

Reply:
Divs are also not allowed in headings.

Reply:
Divs are allowed in headings (in HTML5), they are not allowed in `<p>` (they close it).

Reply:
Although, it can be easily changed to span for inline purposes.

Reply:
> I this case why not wrap the input and the custom markup required in a label? 

@anthonydillon Because parent element can't change style based on child element. And we need label to change when checkbox is ticked or not. The only option is for label to follow the checkbox `input + label`.

Reply:
>> Divs are also not allowed in headings.
>
> Divs are allowed in headings (in HTML5), they are not allowed in <p> (they close it).

@anthonydillon  Or maybe I was mistaken. Validator doesn't seem to accept div in h1, so I'll change it to span and make sure anything else is valid.

Reply:
@anthonydillon @lyubomir-popov I fixed the validation issues.

Replaced the divs with spans everywhere for consistency, properly closed labels.

Reply:
We could add an additional elements which applies the styling now that we controller the markup. As they are all contained in a label they will all be clickable.

Reply:
I only mentioned it because it seems the material UI uses spans probably for the validation reason.

Reply:
@anthonydillon Thanks for comments. If I understand your suggestion correctly it would look something like this?

```html
<label class="p-checkbox" for='check'>
    <span class="p-checkbox__input>
        <input type="checkbox" id='check' />
        <span class="p-checkbox__styled-input"></span>
    </span>
    <span class="p-checkbox__label>I agree to everything</span>
</label>
```

This could work. `<span class="p-checkbox__styled-input">` would render the fake input, it is in DOM after the input so can react to checked/unchecked state. And there is only one label around everything, so technically clicking on anything inside it would affect the input.

It's not much different from current solution in terms of structure (so it should work in terms of porting current styles), but has a benefit of using only one label.

@anthonydillon If you prefer this approach we can park this PR for now and I can get back to it when I'm back from holidays. Better release something solid than rush with quick solution.

Reply:
I would like Lyubos opinion too but I think it's a cleaner approach. I didn't include the extra span in the original example as I thought it could be an `::after` or `::before` element on the text element.

Reply:
OK. Interesting point. We could possibly simplify it even further to:

```html
<label class="p-checkbox">
   <input type="checkbox" />
   <span class="p-checkbox_label>Something</span>
</label>
```

The only 'issue' with that is that this makes it less clear that the `p-checkbox__label` is required and is not just a label, but also a fake input.

The other benefit of what is currently in the PR is that it builds up on what we have in base styles both for checkboxes and labels (like required support). If we change the label to span we will have to reimplement required support (and other label styles) on it for `p-checkbox`. But it's probably worth it looking at how it simplifies the HTML needed.

I would say that it's certainly worth investigating you suggestions @anthonydillon and see how replacing label with span and wrapping everything with label affects this.
I suppose it will mean a bit more complex SCSS, but gives us a nicer HTML which may be a good trade off. So I think it's worth giving it a bit more work.

Reply:
@bartaz the last snippet (label>input+span) seems best to me. unlike the rest, it feels intuitive and easy to remember.